Objects/stringlib/fastsearch.h

```diff
@@ -405,11 +406,18 @@ STRINGLIB(_two_way)(const STRINGLIB_CHAR *haystack, Py_ssize_t len_haystack,
             window = window_last - len_needle + 1;
             assert((window[len_needle - 1] & TABLE_MASK) ==
                    (needle[len_needle - 1] & TABLE_MASK));
+
             Py_ssize_t i = Py_MAX(cut, memory);
             for (; i < len_needle; i++) {
                 if (needle[i] != window[i]) {
-                    LOG("Right half does not match.\n");
-                    window_last += i - cut + 1;
+                    if (i < gap_jump_end) {
+                        LOG("Early right half mismatch: jump by gap.\n");
+                        window_last += gap;
```

Comment:
I'd like to keep the two assertions here and below--it makes it clearer to me that we're just doing `window_last += Py_MAX(gap, i - cut + 1);`.

I think this refactoring makes sense, though I want to do some extra testing to make sure the one extra branch doesn't slow things down in the (common) case where the mismatch is at the first `i`-value.

Reply:
Added them back, but modified to better match `if` clause. Hope it's ok.